pkg/auth/dependency/interaction/provider.go

```diff
@@ -18,6 +18,21 @@ type IdentityProvider interface {
 	New(userID string, typ authn.IdentityType, claims map[string]interface{}) *IdentityInfo
 	CreateAll(userID string, is []*IdentityInfo) error
 	Validate(is []*IdentityInfo) error
+	// RelateIdentityToAuthenticator tells if authenticatorSpec is compatible with and related to identitySpec.
+	//
+	// A authenticatorSpec is compatible with identitySpec if authenticator can be used as authentication for the identity.
+	// For example, OAuth identity is incompatible with any authenticator because the identity itself implicit authenticates.
+	// For example, login ID identity of login ID type username is incompatible with OOB OTP authenticator because
+	// OOB OTP authenticator requires email or phone.
+	//
+	// If authenticatorSpec is incompatible with identitySpec, nil is returned.
+	//
+	// Otherwise authenticatorSpec is further checked if it is related to identitySpec.
+	// If authenticatorSpec is related to identitySpec, authenticatorSpec.Props is mutated in-place.
+	// Currently on the following case mutation would occur.
+	//
+	//   - login ID identity of login ID type email or phone and OOB OTP authenticator.
+	RelateIdentityToAuthenticator(identitySpec IdentitySpec, authenticatorSpec *AuthenticatorSpec) *AuthenticatorSpec
```

Comment:
This function seems to have quite complex semantics. Can we refactor it to be simpler?

For example, it would be simpler to just take the identity and an authenticator type as parameters, and name it as `NewForIdentity`.

Reply:
This function is indeed complex because it is doing three things.

- Check compatible
- Check related
- If related, populate the authenticator spec

It must take an IdentitySpec because it needs the login ID to populate.
It can be refactored to take authenticator type instead because it does not read the claims. It writes only.